lib/engine-universal/src/engine.rs

```diff
@@ -95,6 +95,63 @@ impl UniversalEngine {
         binary: &[u8],
         tunables: &dyn Tunables,
     ) -> Result<crate::UniversalExecutable, CompileError> {
+        // Compute the needed instrumentation
+        struct MaxStackCfg;
+        impl finite_wasm::max_stack::Config for MaxStackCfg {
+            fn size_of_value(&self, ty: finite_wasm::wasmparser::ValType) -> u8 {
+                use finite_wasm::wasmparser::ValType;
+                match ty {
+                    ValType::I32 => 4,
+                    ValType::I64 => 8,
+                    ValType::F32 => 4,
+                    ValType::F64 => 8,
+                    ValType::V128 => 16,
+                    ValType::FuncRef => 8,
+                    ValType::ExternRef => 8, // TODO: is this correct?
+                }
+            }
+            fn size_of_function_activation(
+                &self,
+                locals: &prefix_sum_vec::PrefixSumVec<finite_wasm::wasmparser::ValType, u32>,
+            ) -> u64 {
+                let mut res = 0;
+                res += locals.max_index().map(|l| u64::from(*l).saturating_add(1)).unwrap_or(0) * 8;
+                // TODO: make the above take into account the types of locals by adding an iter on PrefixSumVec that returns (count, type)
+                res += 32; // Rough accounting for rip, rbp and some registers spilled. Not exact.
+                res
+            }
+        }
+        struct GasCfg<'a>(&'a dyn Tunables);
```

Comment:
This probably wants to be inversed I think. That is, it would make a ton of sense to punt on defining this visitor to `nearcore` and have it pass one such as a tunable, rather than wasmer trying to produce a gas config visitor from a tunable.

That way there's an obvious way forward to it starting to charge different amounts of fees depending on instructions, and, more importantly, this visitor is an obvious place to put conditionals based don the protocol version without making wasmer know or care anything about that (yet.)

Reply:
Nice idea! I don’t know why I was so focused on the idea that Tunables must be POD that I didn’t think of it, even though it makes the code much clearer.

However, it seems to me like `wasmparser::VisitOperator` most likely won’t actually be trait-object-able. So it means I’d probably have to make `Tunables` no longer a trait object, and instead a generic parameter throughout the compiler.

It sounds reasonable-ish to me, but I’ll let that thought sink for a few days while I handle other things, and come back to it later; lmk if you have thoughts on the matter :)

Reply:
I use `VisitOperator` (actually a supertrait of it) as a trait object in finite-wasm’s `Module` so it should be okay?